File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    static ArrayList<Integer>[] graph;
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] visited = new int[numCourses];
        graph = new ArrayList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int[] prerequisite : prerequisites) {
            graph[prerequisite[0]].add(prerequisite[1]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (has_cycle(visited, i)) return false;
        }
        return true;
    }

    private static boolean has_cycle(int[] visited, Integer x) {
        if (visited[x] == 1) return true;
        if (visited[x] == 2) return false;
        visited[x] = 1;
        for(Integer node : graph[x]) {
            if (has_cycle(visited, node)) return true;
        }
        visited[x] = 2;
        return false;
    }
}

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [ [] for i in range(numCourses) ]
        visited = [0] * numCourses
    
        for pair in prerequisites:
            graph[pair[1]].append(pair[0])   

        def has_cycle(x: int) -> bool:
            if visited[x] == 1: return True
            if visited[x] == 2: return False
            visited[x] = 1
            for y in graph[x]:
                if has_cycle(y):
                    return True
            visited[x] = 2
            return False

        for i in range(numCourses):
            if has_cycle(i):
                return False

        return True